torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -609,13 +595,24 @@ def alloc_all_gather_outputs(self) -> None:
             alloc_storage(tensor)
 
     def free_unsharded_param(self) -> None:
-        for tensor in itertools.chain(
-            self.all_gather_outputs, self._unsharded_inner_tensors
-        ):
-            free_storage(tensor)
         if ca.compiled_autograd_enabled:
+            """
+            Assumptions under compile:
+            - `self._unsharded_param` is NOT an alias of `self.all_gather_outputs`.
+            Instead, we explicitly *copy* the data from `self.all_gather_outputs`
+            to `self._unsharded_param` in `init_unsharded_param()`.
+            - `self.all_gather_outputs` and `self._unsharded_inner_tensors` are NOT
+            graph inputs. They are created within the graph and is guaranteed to be freed
+            by the end of the graph. They don't leak outside of the graph.
```

Comment:
just to confirm, we don't currently have any optimizations today that try to (potentially unsafely) optimize this `copy_` away?

Reply:
no, we don't have this optimization today

Reply:
discussed offline, it will be helpful for both reducing memory usage and improve perf. Will do this in the next PR.

Reply:
I ended up adding this mentioned pass in this PR - `remove_fsdp2_unsharded_param_graph_input_usage`